### PR TITLE
Handle validation exception on DIG put

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-3535

## Issues
- Validation error on DIG.put returns 500 error

## Changes
- Handle validation exception on DIG put

### CODE REVIEW

**Review by jlhughes — COMMENTED** [review]
Looks good. 
I'll defer to @nanoscott

File: src/foam/nanos/dig/drivers/DigFormatDriver.js [hunk 1]
```diff
@@ -270,7 +270,18 @@ foam.CLASS({
         nu = old.fclone();
         nu.copyFrom(obj);
       }
-      return dao.put(nu);
+
+      try {
+        return dao.put(nu);
+      } catch ( ValidationException ve ) {
+        throw new DAOPutException(ve.getMessage());
+      } catch ( CompoundException ce ) {
```
Comment (kgrgreer) [inline src/foam/nanos/dig/drivers/DigFormatDriver.js:278]: Can you comment this block?

Comment (lchanmann) [reply src/foam/nanos/dig/drivers/DigFormatDriver.js:278]: FObject.validate(x) collects all validation exceptions into a compound exception https://github.com/foam-framework/foam2/blob/aced3a1edf3ec7d8de62f6161b75f37e0cc753a6/src/foam/core/FObject.java#L427-L445 and `ce.getClientRethrowException()` would give the first validation exception from the compound exception.

Comment (kgrgreer) [reply src/foam/nanos/dig/drivers/DigFormatDriver.js:278]: I mean add it as a comment in the code for the benefit of future developers reading the code. Comments here are less useful for them.

File: src/foam/nanos/dig/drivers/DigFormatDriver.js [hunk 2]
```diff
@@ -270,7 +270,18 @@ foam.CLASS({
         nu = old.fclone();
         nu.copyFrom(obj);
       }
-      return dao.put(nu);
+
+      try {
+        return dao.put(nu);
+      } catch ( ValidationException ve ) {
+        throw new DAOPutException(ve.getMessage());
```
Comment (nanoscott) [inline src/foam/nanos/dig/drivers/DigFormatDriver.js]: do you need to main the original exception?
```suggestion
        throw new DAOPutException(ve.getMessage(), ve);
```

File: src/foam/nanos/dig/drivers/DigFormatDriver.js [hunk 3]
```diff
@@ -270,7 +270,18 @@ foam.CLASS({
...
+      } catch ( CompoundException ce ) {
+        var clientEx = ce.getClientRethrowException();
+        if ( clientEx instanceof ValidationException ) {
+          throw new DAOPutException(clientEx.getMessage());
```
Comment (nanoscott) [inline src/foam/nanos/dig/drivers/DigFormatDriver.js]: here too?
```suggestion
          throw new DAOPutException(clientEx.getMessage(), clientEx);
```

File: src/foam/nanos/dig/drivers/DigFormatDriver.js (hunk 2 above)
Comment (lchanmann) [reply src/foam/nanos/dig/drivers/DigFormatDriver.js]: Fixed in https://github.com/foam-framework/foam2/pull/4772/commits/5d03767cfca7ad2e108716f8a94212e953cb0dca.

File: src/foam/nanos/dig/drivers/DigFormatDriver.js (hunk 3 above)
Comment (lchanmann) [reply src/foam/nanos/dig/drivers/DigFormatDriver.js]: Fixed in https://github.com/foam-framework/foam2/pull/4772/commits/5d03767cfca7ad2e108716f8a94212e953cb0dca.

**Review by nanoscott — APPROVED** [review]

File: src/foam/nanos/dig/drivers/DigFormatDriver.js (hunk 1 above)
Comment (lchanmann) [reply src/foam/nanos/dig/drivers/DigFormatDriver.js:278]: Updated in https://github.com/foam-framework/foam2/pull/4772/commits/133dac070ba25f69e7a01f1cd71f8528f9895190.

**Pull request merged**